serverless/wasm/main.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build wasm
+
+// Package main provides a series of entrypoints for using a serverless log from
+// JavaScript in a browser.
+//
+// Build with:
+//   GOOS=js GOARCH=wasm go build -o serverless/wasm/main.wasm -tags wasm ./serverless/wasm/
+//
+// To view locally:
+//   # Copy Go wasm JS shim:
+//   cp "$(go env GOROOT)/misc/wasm/wasm_exec.js" serverless/wasm/
+//   # Run a webserver serving the wasm dir
+//   # e.g. using goexec (to install goexec: go get -u github.com/shurcooL/goexec):
+//   goexec 'http.ListenAndServe(`:8080`, http.FileServer(http.Dir(`serverless/wasm`)))'
+// Then point your browser at the server.
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"syscall/js"
+	"time"
+
+	"github.com/google/trillian-examples/serverless/internal/log"
+	"github.com/google/trillian-examples/serverless/internal/storage"
+	"github.com/google/trillian-examples/serverless/internal/storage/webstorage"
+	"github.com/google/trillian/merkle/rfc6962"
+
+	logfmt "github.com/google/trillian-examples/formats/log"
+)
+
+const (
+	logPrefix = "log"
+	ecosystem = "WASM Log v0"
+)
+
+var (
+	logStorage *webstorage.Storage
+)
+
```

Comment:
Just to note I haven't really reviewed this last file past this point. A quick scan shows nothing that freaks me out, but I think there's mileage in addressing the other comments first to make sure my mental model of this is even in the right ballpark. Reviewing the below for anything detailed requires a context switch into application / user / xss-security type of thinking which I need a rest before doing :-)